examples/dynamic_history_vehicles_replacement.py

```diff
@@ -0,0 +1,212 @@
+
+    def __init__(
+        self,
+        should_trigger: Callable[[Dict[str, Any]], bool],
+        on_trigger: Callable[[Dict[str, Any]], None],
+        context: Dict[str, Any] = {},
+    ) -> None:
+        self._fired = False
+        self._should_trigger = should_trigger
+        self._on_trigger = on_trigger
+        self._context = context
+
+    def update(self, **kwargs):
+        self._context.update(kwargs)
+        if not self._fired and self._should_trigger(self._context):
+            self._on_trigger(self._context)
+            self._fired = True
+
+    @property
+    def context(self):
+        return self._context
+
+
+def main(
+    script: str,
+    scenarios: Sequence[str],
+    headless: bool,
+    seed: int,
+    vehicles_to_replace_randomly: int,
+    episodes: int,
+):
+    assert episodes > 0
+    logger = logging.getLogger(script)
+    logger.setLevel(logging.INFO)
+
+    logger.debug("initializing SMARTS")
+    smarts = SMARTS(
+        agent_interfaces={},
+        traffic_sim=None,
+        envision=None if headless else Envision(),
+    )
+    random_seed(seed)
+
+    scenarios_iterator = Scenario.scenario_variations(scenarios, [])
+    scenario = next(scenarios_iterator)
+
+    for episode in range(episodes):
+        logger.info(f"starting episode {episode}...")
+
+        def should_trigger(ctx: Dict[str, Any]) -> bool:
+            return ctx["elapsed_sim_time"] > 2
+
+        def on_trigger(ctx: Dict[str, Any]):
+            # Define agent specs to be assigned
+            agent_spec = AgentSpec(
+                interface=AgentInterface(waypoints=True, action=ActionSpaceType.Lane),
+                agent_builder=BasicAgent,
+            )
+            if ctx["vehicles_to_replace_randomly"] <= 0:
+                logger.warning(
+                    "default (0) or negative value specified for replacement. Replacing all valid vehicle candidates."
+                )
+                sample = ctx["vehicle_candidates"]
+            else:
+                logger.info(
+                    "Choosing {} vehicles randomly from {} valid vehicle candidates.".format(
+                        ctx["vehicles_to_replace_randomly"],
+                        len(ctx["vehicle_candidates"]),
+                    )
+                )
+                sample = random.sample(
+                    ctx["vehicle_candidates"], ctx["vehicles_to_replace_randomly"]
+                )
+
+            assert len(sample) != 0
+
+            # Select vehicles and map to agent ids & specs
+            vehicles_to_trap = {}
+            for veh_id in sample:
+                agent_id = f"agent-{veh_id}"
+                vehicles_to_trap[veh_id] = (agent_id, agent_spec)
+                ctx["agents"][agent_id] = agent_spec.build_agent()
+
+            # Create missions for selected vehicles
+            veh_missions = scenario.create_dynamic_traffic_history_missions(
+                vehicles_to_trap.keys(), ctx["elapsed_sim_time"]
```

Comment:
Just a note that `vehicles_to_trap.keys()` is `sample`.  (I was wondering why `vehicles_to_trap` needs to be a `dict`, but I think it's fine to leave it that way.)